crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -572,6 +572,71 @@ impl AuthorityStorePruner {
         Ok(())
     }
 
+    async fn prune_executed_tx_digests(
+        perpetual_db: &Arc<AuthorityPerpetualTables>,
+        checkpoint_store: &Arc<CheckpointStore>,
+    ) -> anyhow::Result<()> {
+        let current_epoch = checkpoint_store
+            .get_highest_executed_checkpoint()?
+            .map(|c| c.epoch)
+            .unwrap_or_default();
+
+        if current_epoch < 2 {
+            return Ok(());
+        }
+
+        let target_epoch = current_epoch - 1;
+
+        use sui_types::base_types::TransactionDigest;
+        let start_key = (0u64, TransactionDigest::ZERO);
+        let end_key = (target_epoch, TransactionDigest::ZERO);
+
+        info!(
+            "Pruning executed_transaction_digests for epochs < {} (current epoch: {})",
+            target_epoch, current_epoch
+        );
+
+        let mut batch = perpetual_db.executed_transaction_digests.batch();
+        batch.schedule_delete_range(
+            &perpetual_db.executed_transaction_digests,
+            &start_key,
+            // `to` is non-inclusive so target_epoch and all later epochs are preserved
+            &end_key,
+        )?;
+        batch.write()?;
+        Ok(())
+    }
+
+    #[cfg(tidehunter)]
+    fn prune_executed_tx_digests_th(
+        perpetual_db: &Arc<AuthorityPerpetualTables>,
+        checkpoint_store: &Arc<CheckpointStore>,
+        num_epochs_to_retain: u64,
+    ) -> anyhow::Result<()> {
+        let num_epochs_to_retain = max(num_epochs_to_retain, 1);
+        let current_epoch = checkpoint_store
+            .get_highest_executed_checkpoint()?
+            .map(|c| c.epoch)
+            .unwrap_or_default();
+
+        if current_epoch < num_epochs_to_retain {
+            return Ok(());
+        }
+
+        let target_epoch = current_epoch - num_epochs_to_retain;
+        let last_epoch_to_delete = target_epoch - 1;
```

Comment:
this can underflow if `current_epoch == num_epochs_to_retain`

Reply:
Good catch - I've updated this to ignore the `num_epochs_to_retain` cfg, same as the rocksdb pruning method does